skema/program_analysis/CAST/matlab/matlab_to_cast.py

```diff
@@ -578,73 +569,119 @@ def visit_do_loop_statement(self, node) -> Loop:
-            body_node = self.visit(child)
-            if body_node:
-                if not mi.body:
-                    mi.body = list()
-                mi.body.append(body_node)
-
-        return mi
-
+        def get_case_expression(case_node, identifier):
+            """ return an Operator representing the case test """
+            cell_node = get_first_child_by_type(case_node, "cell")
+            # multiple case arguments
+            if (cell_node):
+                nodes = get_all(cell_node, case_node_types)
+                ast_nodes = valid([self.visit(node) for node in nodes])
+                operand = LiteralValue(
+                    value_type="List",
+                    value = [get_node_value(node) for node in ast_nodes],
+                    source_code_data_type=["matlab", MATLAB_VERSION, "unknown"],
+                    source_refs=[self.node_helper.get_source_ref(cell_node)]
+                )
+                return get_operator("in", [identifier, operand])
+            # single case argument
+            nodes = get_children_by_types(case_node, case_node_types)
+            operand = valid([self.visit(node) for node in nodes])[0]
+            return get_operator("==", [identifier, operand])
+
+        def get_case_body(case_node):
+            """ return the instruction block for the case """
+            block = get_first_child_by_type(case_node, "block")
+            if block:
+                return valid([self.visit(child) for child in block.children])
+            return None
+            
+        def get_model_if(case_node, identifier):
+            """ return conditional logic representing the case """
+            return ModelIf(
+                expr = get_case_expression(case_node, identifier),
+                body = get_case_body(case_node),
+            )
+        
+        # switch statement identifier
+        identifier = self.visit(get_first_child_by_type(node, "identifier"))
+        
+        # n case clauses as 'if then' nodes
+        case_nodes = get_children_by_types(node, ["case_clause"])
+        model_ifs = [get_model_if(node, identifier) for node in case_nodes]
+        for i, model_if in enumerate(model_ifs[1:]):
+            model_ifs[i].orelse = [model_if]
+
+        # otherwise clause as 'else' node after last 'if then' node
+        otherwise_clause = get_first_child_by_type(node, "otherwise_clause")
+        if otherwise_clause:
+            block = get_first_child_by_type(otherwise_clause, "block")
+            if block:
+                last = model_ifs[len(model_ifs)-1]
+                last.orelse = valid([self.visit(child) for child in block.children])
+
+        return model_ifs[0]
 
+    def visit_if_statement(self, node):
+        """ return a node describing if, elseif, else conditional logic"""
+
+        def conditional(conditional_node):
+            """ return a ModelIf struct for the conditional logic node. """
+            ret = ModelIf()
```

Comment:
You'll probably also need to add the source_ref that's missing here too.

Reply:
FIxed